### PR TITLE
Feature: acf field models / data transfer objects

### PR DESCRIPTION
**Pull request by defunctl**
## TLDR;

Automatically transform arrays into objects (Field Models/Data Transfer Objects) and gain automatic type casting, type safe models and portable objects. You never have to remember an array index name again. Directly pass in matching ACF fields and that's it:

```php
// [ 'url' => 'https://tri.be', 'title' => 'Tribe', 'target' => '_self' ]
$link = new \Tribe\Libs\Field_Models\Models\Link( get_field( 'my_link_field' ) );

// These all auto complete for you in your IDE so you no longer have to remember 
// what properties are available for an ACF field that returns an array.
echo $link->url;
echo $link->title;
echo $link->target;
```

We've been using these in a number of projects already but I've had to manually bring them and convert everything, so this should be the last time :smile:.

T3 has a few new projects starting soon and this will save us a lot of time once it's built into the framework from the start.

## What does this do/fix?

- Brings in the [square1-field-models](https://github.com/moderntribe/square1-field-models) library! These are built on top of the [Spatie DTO v2 library](https://github.com/spatie/data-transfer-object/tree/v2) (you should check the README).
- Adds custom Field Models/Collections for blocks and components.
- Updates WordPress to 5.9.3, ACF to 5.12.2 and Yoast to 18.8.
- Hides deprecation notices when running lefthook phpcs, which appear if you're running PHP8.1 locally.
- Converts all blocks/components to use Field Models, where appropriate.
- General code clean up, removal of constants, variables and methods that aren't in use + escaping.

## Why Field Models / DTO's?

1. Field models/DTO's remove the inconsistent typing from ACF and allows you populate a Field Model with a single `get_field()` call. You'll always get the expected type back from a Field Model, e.g. [$link->url](https://github.com/moderntribe/square1-field-models/blob/main/src/Models/Link.php#L10) will always be a string.
2. IDE code completion. Set a Field Model once and pass around the instance you'll see all available properties without having to remember what an ACF field consists of. [Do you know all the ACF File Field indexes and their types?](https://github.com/moderntribe/square1-field-models/blob/main/src/Models/File.php#L7)
3. Simpler and much less code (_see before and after below_). Fewer model/controller constants and because the Field Model travels through these instances, you only have to add a new property to the Field Model and then you can use it in the controller or view without adding additional controller and model constants/properties. 
5. Collections let you create many of the same Field Model, with full IDE completion, assuming you type hint the collection you made. [example](https://github.com/moderntribe/square-one/blob/eccf61524df04437be4d4055dc2faefc8d976a9c/wp-content/plugins/core/src/Templates/Models/Collections/Logo_Collection.php). Collections are extremely memory efficient, as they are iterators out of the box, so they can hold an extreme amount of data, and will use minimal memory when `foreach()` looping over them. They can be converted to an array, which will utilize full memory when absolutely necessary. [Read more about collections](https://github.com/spatie/data-transfer-object/tree/v2#working-with-collections).
6. It's easy to make your own Field Models and Collections. Just extend the [Field_Model class](https://github.com/moderntribe/square1-field-models/blob/main/src/Field_Model.php), add typed properties that match your ACF field names and your data will get auto-populated based on what `get_field()` returns. 

[These are the ACF models I've made so far](https://github.com/moderntribe/square1-field-models/tree/main/src/Models) out of the box for ACF, but any ACF field that returns an array can be made into one, so if you have any ideas, please submit a PR. If you want to know how things work under the hood, I think the [integration tests](https://github.com/moderntribe/square1-field-models/tree/main/tests/integration/Tribe/Libs/Field_Models) do a good job of that.

> You can bring the ACF models and DTO library into an older project if you wish, as long as it's PHP7.4+ with `so composer require moderntribe/square1-field-models`.

---

## Code Examples

Some before and afters showing reduced code required.

### Before: CTA

```php
// compacted block model example
class Stats_Model extends Base_Model {

	public function get_data(): array {
		return [
			Stats_Block_Controller::ATTRS => $this->get_attrs(),
			Stats_Block_Controller::CTA   => $this->get_cta_args(),
		];
	}

	private function get_cta_args(): array {
		$cta  = $this->get( Cta_Field::GROUP_CTA, [] );
		$link = wp_parse_args( $cta['link'] ?? [], [
			'title'  => '',
			'url'    => '',
			'target' => '',
		] );

		return [
			Link_Controller::CONTENT        => $link['title'],
			Link_Controller::URL            => $link['url'],
			Link_Controller::TARGET         => $link['target'],
			Link_Controller::ADD_ARIA_LABEL => $cta['add_aria_label'] ?? false,
			Link_Controller::ARIA_LABEL     => $cta['aria_label'] ?? '',
		];
	}
}

// compacted block controller
class Stats_Block_Controller extends Abstract_Controller {

	public const CTA = 'cta';

	/**
	 * @var string[]
	 */
	private array $cta;

	public function __construct( array $args = [] ) {
		$args      = $this->parse_args( $args );
		$this->cta = (array) $args[ self::CTA ];
	}

	public function get_header_args(): array {
		return [
			Content_Block_Controller::TAG => 'header',
			Content_Block_Controller::CTA => $this->get_cta(),
		];
	}

	protected function defaults(): array {
		return [
			self::CTA => [],
		];
	}

	private function get_cta(): Deferred_Component {
		$cta = wp_parse_args( $this->cta, [
			'content'        => '',
			'url'            => '',
			'target'         => '',
			'add_aria_label' => false,
			'aria_label'     => '',
		] );

		return defer_template_part( 'components/link/link', null, [
			Link_Controller::URL            => $cta['url'],
			Link_Controller::CONTENT        => $cta['content'] ?: $cta['url'],
			Link_Controller::TARGET         => $cta['target'],
			Link_Controller::ADD_ARIA_LABEL => $cta['add_aria_label'],
			Link_Controller::ARIA_LABEL     => $cta['aria_label'],
			Link_Controller::CLASSES        => [
				'c-block__cta-link',
				'a-btn',
				'a-btn--has-icon-after',
				'icon-arrow-right',
			],
		] );
	}

}

```

### After: CTA

```php
// compacted block model example
class Stats_Model extends Base_Model {

	public function get_data(): array {
		return [
			Stats_Block_Controller::ATTRS => $this->get_attrs(),
			Stats_Block_Controller::CTA   => new Cta( $this->get( Cta_Field::GROUP_CTA, [] ) ),
		];
	}

}

// compacted block controller
class Stats_Block_Controller extends Abstract_Controller {

	public const CTA = 'cta';

	private Cta $cta;

	public function __construct( array $args = [] ) {
		$args      = $this->parse_args( $args );
		$this->cta = $args[ self::CTA ];
	}

	public function get_header_args(): array {
		return [
			Content_Block_Controller::TAG => 'header',
			Content_Block_Controller::CTA => $this->get_cta(),
		];
	}

	protected function defaults(): array {
		return [
			self::CTA => new Cta(),
		];
	}

	private function get_cta(): Deferred_Component {
		return defer_template_part( 'components/link/link', null, [
			Link_Controller::URL            => $this->cta->link->url,
			Link_Controller::CONTENT        => $this->cta->link->title ?: $this->cta->link->url,
			Link_Controller::TARGET         => $this->cta->link->target,
			Link_Controller::ADD_ARIA_LABEL => $this->cta->add_aria_label,
			Link_Controller::ARIA_LABEL     => $this->cta->aria_label,
			Link_Controller::CLASSES        => [
				'c-block__cta-link',
				'a-btn',
				'a-btn--has-icon-after',
				'icon-arrow-right',
			],
		] );
	}

}
```

Take an [ACF Link Field](https://www.advancedcustomfields.com/resources/link/) as an example. If not populated, `get_field( 'my_link_field' );` can return null, false, an empty array etc...

Take our [Link Field Model](https://github.com/moderntribe/square1-field-models/blob/main/src/Models/Link.php) and just pass in the field output:

```php
// get_field() outputs this data: [ 'url' => 'https://tri.be', 'title' => 'Tribe', 'target' => '_self' ]
$link = new \Tribe\Libs\Field_Models\Models\Link( get_field( 'my_link_field' ) );

// Need an array, try...
$link->toArray(); // [ 'url' => 'https://tri.be', 'title' => 'Tribe', 'target' => '_self' ]

// Need only some of the field data?
$link->only( 'url' )->toArray();  // [ 'url' => 'https://tri.be' ]
$link->except( 'target' )->toArray();  // [ 'url' => 'https://tri.be', 'title' => 'Tribe' ]
```

## QA

- It's important to note field changes here are not backwards compatible (renaming ACF fields means the data no longer renders). As SquareOne is forked from its current state for new projects, this doesn't matter but would if you were testing with an existing database.
- Blocks tested: http://p.tri.be/Ayokf8
- Everything should work exactly as before, but with less code and more code completion!

## Tests

Does this have tests?

- [ ] Yes
- [x] No, this doesn't need tests because all tests are in the [square1-field-models repo](https://github.com/moderntribe/square1-field-models)
- [ ] No, I need help figuring out how to write the tests.

### CODE REVIEW

**Review by bjcooper — APPROVED** [review]
This is awesome, Justin! Having used it on a few projects already more-or-less, the code completion and lack of magic strings is a huge step up.

**Review by verytwisty — APPROVED** [review]
Great work - looks a lot simpler to use :-D

**Review by ScottGarrison — APPROVED** [review]

**Review by JMRhodes — APPROVED** [review]
🙌

**Review by ChrisMKindred — APPROVED** [review]
This looks good and the IDE completion is extremely helpful.

I'd like to add some info about this to the doc folder and make a quick pass to verify that we aren't referencing any of the "old" ways of handling the fields to help eliminate any confusion.

**Review by dpellenwood — APPROVED** [review]
Love this Justin! Thanks for tackling it.

**Review by nickpelton — APPROVED** [review]
Looked over the general pattern and I think this is a great step. I had a few thoughts along the way:

1. Some primitive components like links in a CTA, I often it passing a CTA and then breaking the CTA up for the usage for a link component. I do wonder if there is an argument if we should be building models more explicitly for each component and just passing them directly rather than needing to break them up as we pass to the component.  It's a future debate.

2.  Now that building components is going to require even more steps in a few cases, we should see if we need to update the generator to accommodate or documentation as well.

**Review by jamiepaul — APPROVED** [review]
🖤

File: wp-content/themes/core/components/blocks/buttons/buttons.php [hunk 1]
```diff
@@ -1,19 +1,36 @@
...
-	<?php } ?>
+	<?php foreach ( $c->get_buttons() as $button ) {
+		if ( ! $button->cta->link->url ) {
+			continue;
+		}
+
+		$style = $button->button_style ? sanitize_html_class( sprintf( 'b-style__%s', $button->button_style ) ) : '';
+
+		get_template_part( 'components/link/link', '', [
+			Link_Controller::URL            => $button->cta->link->url,
```
[the remark this answers is not in the document]
Comment (defunctl) [reply wp-content/themes/core/components/blocks/buttons/buttons.php:25]: This is interesting, obviously it would then couple the Field Model implementation to the component and make things less modular. On the other hand, some kind of `Link_Controller::create_from_cta( Cta $cta ): array;` or something like that could be useful, but that the end of the day I believe the CSS classes on all of these are different based on the component they are used in.

File: wp-content/themes/core/components/blocks/content_loop/Content_Loop_Controller.php [hunk 2]
```diff
@@ -232,24 +230,13 @@ public function get_header_args(): array {
...
-			'aria_label'     => '',
-		] );
-
 		return defer_template_part( 'components/link/link', null, [
-			Link_Controller::URL            => $cta['url'],
-			Link_Controller::CONTENT        => $cta['content'] ?: $cta['url'],
-			Link_Controller::TARGET         => $cta['target'],
-			Link_Controller::ADD_ARIA_LABEL => $cta['add_aria_label'],
-			Link_Controller::ARIA_LABEL     => $cta['aria_label'],
+			Link_Controller::URL            => $this->cta->link->url,
```
[the remark this answers is not in the document]
Comment (defunctl) [reply wp-content/themes/core/components/blocks/content_loop/Content_Loop_Controller.php:235]: Same thing, the CSS classes are what is making each one different in this case. 

If we centralize the component building of this, there's a chance someone makes a change and breaks all the other blocks using it. 

I'm sure there is a way to do this more elegantly though, but at some point some central factory or creator will just get more and more things added to it as the CTA differentiates across each component, making that central call more and more risky. 

I guess it's a trade off between risk of regressions vs code duplication on this one, but personally I think if you make a change to the rendering of the CTA component in a controller, it should only affect that controller.

Would love to hear more though, maybe I'm looking at this in the wrong way.

[the remark this answers is not in the document]
Comment (nickpelton) [reply wp-content/themes/core/components/blocks/content_loop/Content_Loop_Controller.php:235]: I think you are thinking the right way, I think maybe the components could now be optimized for this approach. Like could the CTA model have a Link Model property and simply accept a link model but also a low for overrides?

```php
get_template_part( 'components/link/link', '', [
'link_model' => $link_model,
'classes' => 'classes'
]
```

**Review by alexander-kozhukhovski — APPROVED** [review]
Looks great! 🌮 🌮 🌮

**Review by lindseyolson — APPROVED** [review]
